src/HexagonOptimize.cpp

```diff
@@ -162,6 +162,8 @@ std::vector<Pattern> casts = {
     // Saturating narrowing casts
     { "halide.hexagon.satub.shr.vh.h", u8c(wild_i16x/wild_i16), Pattern::DeinterleaveOps | Pattern::ExactLog2Op1 },
     { "halide.hexagon.satuh.shr.vw.w", u16c(wild_i32x/wild_i32), Pattern::DeinterleaveOps | Pattern::ExactLog2Op1 },
+    { "halide.hexagon.sath.shr.vw.w", i16c(wild_i32x/wild_i32), Pattern::DeinterleaveOps | Pattern::ExactLog2Op1 },
+    { "halide.hexagon.satub.shr.vw.w", u8c(wild_i32x/wild_i32), Pattern::DeinterleaveOps | Pattern::ExactLog2Op1 },
```

Comment:
That makes sense, but I think it's problematic, because it's ambiguous whether DeinterleaveOps means that two levels of deinterleavings should happen, or just one. In other words, how does CodeGen_Hexagon disambiguate between deinterleaving a quad vector of words once (as would happen if you vectorized a narrowing 2x the natural vector size of the narrow type) and deinterleaving a quad word that is the input to a 2x narrowing? I think both will look like a call to deinterleave.vw.dv.

I think the right way to express this is to get it so that there are two deinterleave calls composed with eachother in the IR, the "inner" one being a deinterleave.vw, and the outer one being a deinterleave.vh.

We can't currently express this with the pattern system we have now. It might be best to just add some special case code in OptimizePatterns::visit(const Cast*) to handle these double narrowings.

Another advantage of generating two nested deinterleave calls is that the interleaving cancellation should simplify away one (if the input data was widened to begin with) or both (if the input data was double widened to begin with). If you use a baked in double deinterleaving, we'd have to modify the interleave cancellation to understand that, which would be much harder than just generating nested deinterleave calls (some double deinterleavings would get replaced by a single deinterleaving...).